### PR TITLE
Wave3: Prodigy Adaptive Learning Rate Optimizer — cross-dataset

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Prodigy is a parameter-free adaptive optimizer that automatically estimates the optimal learning rate during training, eliminating the need for LR tuning. It extends D-Adaptation with bias correction. In our setting, where optimal LRs differ significantly across datasets (Lion 1e-4 for TF, AdamW 6e-4 for AF, AdamW 5e-4 for DM), a single auto-tuning optimizer could enable a unified cross-dataset config and potentially find better solutions than our hand-tuned baselines.

## Prodigy Setup

Install: `pip install prodigyopt`

```python
from prodigyopt import Prodigy

optimizer = Prodigy(
    model.parameters(),
    lr=1.0,                    # Prodigy rescales this automatically
    weight_decay=1e-2,
    safeguard_warmup=True,     # Prevents LR explosion in warmup phase
    use_bias_correction=True,  # Recommended for transformers
    d_coef=1.0,                # Default; try 0.1 if diverging
)
```

Add `--optimizer prodigy` flag. Keep `--weight-decay` unchanged. The `--lr` flag should be set to 1.0 (Prodigy's convention) but this can be overridden.

**Important**: Prodigy is incompatible with cosine annealing schedulers in some implementations. Use a constant LR schedule (no `--cosine-t-max`) first. If training is stable, optionally add cosine warmup only.

## Training Instructions — All 4 Datasets

Use `--wandb_group gilbert-prodigy`. Run one trial per dataset.

### TandemFoil
```bash
cd target/icml2026

python train.py --dataset tandemfoil --optimizer prodigy --lr 1.0 \
  --weight-decay 1e-2 --no-use-ema \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition --epochs 999 \
  --wandb_group gilbert-prodigy
```
Note: omit `--cosine-t-max` and `--grad-clip` initially (Prodigy handles LR adaptation). If diverging, add `--grad-clip 1.0`.

### TandemFoil Paper (use EMA)
```bash
python train.py --dataset tandemfoil_paper --optimizer prodigy --lr 1.0 \
  --weight-decay 1e-2 --ema-decay 0.999 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition --epochs 999 \
  --wandb_group gilbert-prodigy
```

### AirfRANS
```bash
python train.py --dataset airfrans --airfrans-task full --optimizer prodigy \
  --lr 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 \
  --model-heads 4 --epochs 999 \
  --wandb_group gilbert-prodigy
```

### DrivAerML
```bash
SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer prodigy --lr 1.0 \
  --weight-decay 1e-2 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group gilbert-prodigy
```

## Troubleshooting

- If any run diverges early, add `--grad-clip 1.0` and restart
- If Prodigy's estimated LR (logged as `dlr` or `d*lr` in W&B) is very small (<1e-6) or very large (>0.1), try `--d_coef 0.1`
- If cosine annealing is needed for convergence, re-run with `--cosine-t-max 10`

## Current Baselines (targets to beat)

| Dataset | Metric | Current Best | PR |
|---------|--------|-------------|-----|
| TandemFoil | `val_primary/surface_pressure_mae` | **26.06** | #2887 (gohan) |
| TandemFoil Paper | `val_primary/field_mse` | TBD (establishing via #2947-2949) |
| AirfRANS | `val_primary/surface_mse` | **0.000627** | #2902 (stark) |
| DrivAerML | `val_primary/surface_rel_l2_pct` | **3.997%** | #2898 (piccolo) |

## Expected Outcome

Prodigy's auto-tuned LR could match or exceed our hand-tuned LRs. The key risk is instability in early training — `safeguard_warmup=True` mitigates this. Report the estimated LR (d*lr) from W&B alongside the primary validation metrics for all 4 datasets.

### CODE REVIEW

**Comment by morganmcg1** [comment]
Closing to narrow the queue for the final ICML benchmark sprint. Current priority is benchmark-specific optimization and benchmark-faithful reporting for DrivAerML, TandemFoil paper benchmark, AirfRANS full Surf/Vol, and the Tandem parity anchor. If this idea still matters later, we can reopen it in a benchmark-specific form.
